### PR TITLE
handle unexpected wakeup on kobo

### PR DESCRIPTION
**Pull request by houqp**
Putting the code out here for early review. I am still testing this myself and please help test it too if you can :)

For https://github.com/koreader/koreader/issues/2520 and https://github.com/koreader/koreader/issues/2239

### CODE REVIEW

**Comment by Frenzie** [comment]
I like the new kodev command.

File: frontend/device/kobo/device.lua [hunk 1]
```diff
@@ -320,8 +320,25 @@ function Kobo:getFirmwareVersion()
...
+               unexpected_wakeup_count)
+    unexpected_wakeup_count = unexpected_wakeup_count + 1
+    Kobo.suspend()
+end
+
+function Kobo:getUnexpectedWakeup() return unexpected_wakeup_count end
+
 function Kobo:suspend()
-    logger.info("Kobo Suspend: Going to sleep . . .")
+    logger.info("Kobo suspend: Going to sleep . . .")
```
Comment (Frenzie) [inline frontend/device/kobo/device.lua]: If you're getting rid of caps I'd also drop those after the colon.

File: kodev [hunk 2]
```diff
@@ -319,6 +319,63 @@ OPTIONS:
...
+
+    check_submodules && make
+    setup_env
+    make "${EMU_DIR}/.busted"
+    pushd "${EMU_DIR}"
+        target=front
+        test_path="./spec/${target}/unit"
+        if [ ${show_previous} -eq 0 ]; then
+            echo "Runing tests in" ${test_path}
+            busted --lua="./luajit" ${opts} \
```
Comment (Frenzie) [inline kodev]: Should `${opts}` be quoted?

File: kodev [hunk 3]
```diff
@@ -319,6 +319,63 @@ OPTIONS:
...
+    done
+
+    check_submodules && make
+    setup_env
+    make "${EMU_DIR}/.busted"
+    pushd "${EMU_DIR}"
+        target=front
+        test_path="./spec/${target}/unit"
+        if [ ${show_previous} -eq 0 ]; then
+            echo "Runing tests in" ${test_path}
```
Comment (Frenzie) [inline kodev]: Running.

**Comment by houqp** [comment]
Done, seems to be working fine for me in the sense that it did not break the existing suspend behavior. I have never run into unexpected wakeup myself, so I can't really test this part.

**Comment by poire-z** [comment]
You may try to simulate spurious wakeups by launching a backgrounded shell script with stuff like:
```
sleep 234
echo +456 > /sys/class/rtc/rtc0/wakealarm
sleep 123
echo +789 > /sys/class/rtc/rtc0/wakealarm
```
This schedules a wakeup in 456 or 789 seconds.

Dunno if there's a chance for kobo linux stuff or nickel to do this kind of programming of wakeups, but if we would want to cancel them, we could use at start time (or just before going to suspend) `echo 0 > /sys/class/rtc/rtc0/wakealarm` to remove any scheduled one left over by nickel.

**Comment by houqp** [comment]
Interesting, I will give this a try :)

**Comment by Frenzie** [comment]
Um, so I did squash and merge (into "one commit in the base branch") on #2445 and it ended up here. Even though `master` is the default/base branch and not houqp-master. I guess it can hang out in this PR but that's pretty weird. >_>

**Comment by poire-z** [comment]
Forgot to report that i had this patch on my gloHD for 2-3 weeks, with no side effect encountered.

**Comment by Frenzie** [comment]
@houqp How about it? Is it ready? /peer pressure

**Comment by houqp** [comment]
Not ready yet, I noticed a potential issue with putting the device directly back to suspend without running through the resume code path. I need to double check on this.

**Comment by houqp** [comment]
@Frenzie this one is ready for merge

**Comment by poire-z** [comment]
@Frenzie : just pinging you to the [above trick](https://github.com/koreader/koreader/pull/2527#issuecomment-277501777) to have something like wakealarm working (it does on my 3.19 GloHD).

**Comment by Frenzie** [comment]
@poire-z I linked that principle [here](https://www.mobileread.com/forums/showthread.php?p=3815045#post3815045) (after an edit), but on my H2O I get this error:
```
-sh: can't create /sys/class/rtc/rtc0/wakealarm: nonexistent directory 
```

Anyway, it's purely of academic interest to me. On my laptop it might make sense to have it hibernate after a few hours of standby since otherwise it'll drain itself within… I don't know, a week? But my H2O takes a couple of months. I suppose the auto-shutdown principle could still be interesting to save battery, but I'd set it to at least a few days. Anything shorter would probably not even have all that much in the way of perceptible savings, with all the processing & drawing involved in shutting down and cold starting.

**Pull request merged**